keps/sig-network/3015-node-level-topology/README.md

```diff
@@ -0,0 +1,436 @@
+outside the cluster to be delivered to my pods as efficiently as
+possible (as with `Local` traffic policy), but I don't care if the
+source IP is preserved, and I _do_ care about maximizing reliability
+when pods restart or move between nodes (which can cause hiccups with
+`externalTrafficPolicy: Local` in some cases).
+
+#### Per-Node DNS Pods
+
+As a cluster administrator, I plan to run a DNS pod on each node, and
+would like DNS requests from other pods to always go to the local DNS
+pod, for efficiency. However, if no local DNS pod is available, DNS
+should automatically just go to a remote pod instead.
+
+(This is a modified version of a user story from the [Internal Traffic
+Policy KEP]. The original implies (by omission) that is is acceptable
+to have DNS lookups fail when there is no local DNS pod.)
+
+[Internal Traffic Policy KEP]: ../2086-service-internal-traffic-policy/
+
+## Design Details
+
+### Applying Node-Level Topology
+
+Given a Service to which node-level topology applies (see next
+section), the implementation is simple: if there is at least one
+endpoint for the service on the local node, then filter the list of
+endpoints to include only local ones. (If there are no local
+endpoints, then use the same rules as would have been used if
+node-level topology was not in effect.)
+
+As with Topology Aware Hints, Node-Level Topology would only apply to
+connections with `Cluster` traffic policy, because
+`internalTrafficPolicy: Local` semantically _requires_ local delivery
+(and is not allowed to fall back when there are no local endpoints),
+and `externalTrafficPolicy: Local` requires source IP preservation
+(which is generally not possible when sending to a non-local
+endpoint).
+
+### Choosing Services to Apply Node-Level Topology To
+
+```
+<<[UNRESOLVED]>>
+
+Pick an approach.
+
+The Automatic approach seems simplest (and would have the effect of
+immediately making DNS faster in most(?) clusters). But maybe it would
+do bad things in some circumstances?
+
+<<[/UNRESOLVED]>>
+```
+
+#### Approach 1: Manual
+
+One approach would be to allow users to explicitly tag services where
+they want this behavior.
+
+This could be implemented via the existing
+`"service.kubernetes.io/topology-aware-hints"` annotation perhaps.
+Instead of setting it to `Auto`, the user could set it to `Node` to
+indicate node-level topology rather than zone-level.
+
+(This would imply that a service could not use both node-level
+topology and zone-level topology. Another possibility would be to have
+a new annotation for node-level topology.)
+
+#### Approach 2: Automatic
+
+Alternatively, the EndpointSlice controller could attempt to determine
+automatically which services would benefit from node-level topology.
+For example, the EndpointSlice controller could look for services
+where:
+
+  - The service has an endpoint on every node, or at least "almost
+    every" node. (eg, no more than N or N% of nodes are missing an
+    endpoint).
```

Comment:
What if every node has multiple endpoints, but the distribution is uneven?

Reply:
Do we believe that this is likely to actually happen?

If we think this is a _common_ case then we could adjust the rules for it (eg, in this case make the service be biased toward using local endpoints, but not exclusively so).

There is definitely a difference between "a Service which is intentionally run on every node" and "a Service which happens to have more endpoints than the cluster has nodes". Note that if we used the proposed "DaemonSet" rule, then we would just not use node-level topology in this situation. (Which could be good or bad... ?)

Reply:
Agree that there's a significant difference here, but in the case where you have more endpoints than nodes, why wouldn't you want this feature enabled? If many/most of the current uses of ExternalTrafficPolicy=Local are to avoid a second hop, wouldn't this just be a safer way to do that?

Reply:
Node-level topology doesn't really make sense with services that aren't actually keyed to the number of nodes (ie, DaemonSets).

Like, if you have 10 nodes, and you have some service whose pods are created by a Deployment, and you decide that you're going to need 10 replicas in order to process all of the incoming requests, then sure, if those 10 replicas end up on the 10 separate nodes, then automatically applying node-level topology would probably make things work faster...

But if the service starts to get overloaded and laggy, and you decide you need to scale it up to 12 replicas to handle the load, what then? If we keep using node-level topology, then you'd end up with 2 nodes with a pair of under-utilized endpoints, and 8 nodes with a single still-overloaded endpoint. You wouldn't actually manage to address all of the overloading until you scaled all the way up to 20 replicas, which is way more than you actually needed...